### PR TITLE
orca-slicer: fix webkit2gtk

### LINKED ISSUE #358666: orca-slicer 2.2.0 build fails on master due to missing webkit2gtk-4.1
## Describe the bug

Currently, orca-slicer build fails on master:

```console
error: builder for '/nix/store/1q4nrwy5zr4xkpyaxkg88jnw1kmfyhib-orca-slicer-2.2.0.drv' failed with exit code 1;
       last 25 log lines:
       > -- Looking for XOpenDisplay in /nix/store/qrj5f1mb08k5jg7y9ikzb1njli0gvmxz-libX11-1.8.10/lib/libX11.so;/nix/store/hpd1dss7lmrywlr37vncyzbhzp6rgri5-libXext-1.3.6/lib/libXext.so
       > -- Looking for XOpenDisplay in /nix/store/qrj5f1mb08k5jg7y9ikzb1njli0gvmxz-libX11-1.8.10/lib/libX11.so;/nix/store/hpd1dss7lmrywlr37vncyzbhzp6rgri5-libXext-1.3.6/lib/libXext.so - found
       > -- Looking for gethostbyname
       > -- Looking for gethostbyname - found
       > -- Looking for connect
       > -- Looking for connect - found
       > -- Looking for remove
       > -- Looking for remove - found
       > -- Looking for shmat
       > -- Looking for shmat - found
       > -- Looking for IceConnectionNumber in ICE
       > -- Looking for IceConnectionNumber in ICE - found
       > -- Checking for module 'webkit2gtk-4.1'
       > --   No package 'webkit2gtk-4.1' found
       > CMake Error at /nix/store/4kb60dy0dg80ww8gh2av4by4vwn8l48k-cmake-3.30.5/share/cmake-3.30/Modules/FindPkgConfig.cmake:645 (message):
       >   The following required packages were not found:
       >
       >    - webkit2gtk-4.1
       >
       > Call Stack (most recent call first):
       >   /nix/store/4kb60dy0dg80ww8gh2av4by4vwn8l48k-cmake-3.30.5/share/cmake-3.30/Modules/FindPkgConfig.cmake:873 (_pkg_check_modules_internal)
       >   src/slic3r/CMakeLists.txt:666 (pkg_check_modules)
       >
       >
       > -- Configuring incomplete, errors occurred!
       For full logs, run 'nix log /nix/store/1q4nrwy5zr4xkpyaxkg88jnw1kmfyhib-orca-slicer-2.2.0.drv'.
```

## Steps To Reproduce

1. use master branch as your nixpkgs input
2. add `orca-slicer` to your packages
3. `nixos-rebuild`

## Expected behavior

I expect the build to succeed.

## Screenshots

N/A

## Additional context

No additional context.

## Metadata

```console
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.62, NixOS, 25.05 (Warbler), 25.05.20241124.30abd65`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.10`
 - nixpkgs: `/etc/nixpkgs/channels/nixpkgs`
```

## Notify maintainers

@zhaofengli
@ovlach
@pinpox

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by liberodark**
Fix : https://github.com/NixOS/nixpkgs/issues/358666

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by liberodark** [comment]
Hi,

Also this bug need to be backport to 24.11
Best Regards

**Comment by pinpox** [comment]
@ofborg build orca-slicer bambu-studio

File: pkgs/by-name/or/orca-slicer/package.nix [hunk 1]
```diff
@@ -15,6 +17,18 @@ bambu-studio.overrideAttrs (
       hash = "sha256-h+cHWhrp894KEbb3ic2N4fNTn13WlOSYoMsaof0RvRI=";
     };
 
+    nativeBuildInputs = (previousAttrs.nativeBuildInputs or [ ]) ++ [
+      pkg-config
+    ];
+
+    buildInputs = (previousAttrs.buildInputs or [ ]) ++ [
```
Comment (Scrumplex) [inline pkgs/by-name/or/orca-slicer/package.nix:24]: Let's remove webkitgtk_4_0 to reduce the build closure.


```suggestion
    buildInputs = (lib.remove webkitgtk_4_0 previousAttrs.buildInputs or [ ]) ++ [
```

**Review by Scrumplex — COMMENTED** [review]
I feel like we should separate this package from bambu-studio in the future. Overrides get difficult to maintain the more the projects diverge.

**Comment by Scrumplex** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 358948`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>orca-slicer</li>
    <li>orca-slicer.debug</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 358948`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>orca-slicer</li>
    <li>orca-slicer.debug</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>orca-slicer</li>
    <li>orca-slicer.debug</li>
  </ul>
</details>

**Comment by liberodark** [comment]
> I feel like we should separate this package from bambu-studio in the future. Overrides get difficult to maintain the more the projects diverge.

Yes I think it should be separated.
I can do it if necessary.

File: pkgs/by-name/or/orca-slicer/package.nix (hunk 1 above)
Comment (liberodark) [reply pkgs/by-name/or/orca-slicer/package.nix]: Have try to apply your suggestion but that not help.

Comment (Scrumplex) [reply pkgs/by-name/or/orca-slicer/package.nix:24]: The reason the build fails is that the patch needs to be updated to link against webkit2gtk-4.1.

I was able to build orca-slicer with the following patch:

```patch
From 414dfb6cef6cb144625982a02296bf3086e8431e Mon Sep 17 00:00:00 2001
From: Sefa Eyeoglu <contact@scrumplex.net>
Date: Wed, 27 Nov 2024 19:53:32 +0100
Subject: [PATCH] orca-slicer: link against webkitgtk-4.1

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>
---
 pkgs/by-name/or/orca-slicer/package.nix                        | 3 ++-
 ...-not-for-upstream-CMakeLists-Link-against-webkit2gtk-.patch | 2 +-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/pkgs/by-name/or/orca-slicer/package.nix b/pkgs/by-name/or/orca-slicer/package.nix
index 00c3cdea2938..9ff5e6a1e490 100644
--- a/pkgs/by-name/or/orca-slicer/package.nix
+++ b/pkgs/by-name/or/orca-slicer/package.nix
@@ -2,6 +2,7 @@
   lib,
   fetchFromGitHub,
   bambu-studio,
+  webkitgtk_4_0,
   webkitgtk_4_1,
   pkg-config,
 }:
@@ -21,7 +22,7 @@ bambu-studio.overrideAttrs (
       pkg-config
     ];
 
-    buildInputs = (previousAttrs.buildInputs or [ ]) ++ [
+    buildInputs = (lib.remove webkitgtk_4_0 previousAttrs.buildInputs or [ ]) ++ [
       webkitgtk_4_1
     ];
 
diff --git a/pkgs/by-name/or/orca-slicer/patches/0001-not-for-upstream-CMakeLists-Link-against-webkit2gtk-.patch b/pkgs/by-name/or/orca-slicer/patches/0001-not-for-upstream-CMakeLists-Link-against-webkit2gtk-.patch
index 15f1bf8f0b59..4497271938b5 100644
--- a/pkgs/by-name/or/orca-slicer/patches/0001-not-for-upstream-CMakeLists-Link-against-webkit2gtk-.patch
+++ b/pkgs/by-name/or/orca-slicer/patches/0001-not-for-upstream-CMakeLists-Link-against-webkit2gtk-.patch
@@ -23,7 +23,7 @@ index 9c5cb96..e92a0e3 100644
  
 +# We link against webkit2gtk symbols in src/slic3r/GUI/Widgets/WebView.cpp
 +if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
-+    target_link_libraries(libslic3r_gui "-lwebkit2gtk-4.0")
++    target_link_libraries(libslic3r_gui "-lwebkit2gtk-4.1")
 +endif ()
 +
  # Link the resources dir to where Slic3r GUI expects it
-- 
2.47.0
```

**Comment by liberodark** [comment]
Other issue is : 

```
./result/bin/orca-slicer
(process:217297): libsoup-ERROR **: 08:37:22.425: libsoup3 symbols detected. Using libsoup2 and libsoup3 in the same process is not supported.
fish: Job 1, './result/bin/orca-slicer' terminated by signal SIGTRAP
```

```
ldd ./result/bin/.orca-slicer-wrapped | grep -i soup
    libsoup-3.0.so.0 => /nix/store/vmsggid85874ml3nhjz07i2dz30qjs3h-libsoup-3.6.0/lib/libsoup-3.0.so.0 (0x00007f357f423000)
    libsoup-2.4.so.1 => /nix/store/wqjfrwyl7y1g8vhlkj454xm9kaljipxf-libsoup-2.74.3/lib/libsoup-2.4.so.1 (0x00007f3579c7a000)
```

**Comment by Scrumplex** [comment]
This is gonna require a lot of changes. We need to build wxGTK with webkitgtk_4_1 support. Upstream has a fork of wxWidgets with many changes to support this in their own builds: https://github.com/SoftFever/Orca-deps-wxWidgets

That would involve backporting https://github.com/wxWidgets/wxWidgets/pull/23633. I am not sure if we want to do this for everyone, or just Orca, but considering it has been merged upstream, it *might* be okay to backport it in general.

There is also https://github.com/wxWidgets/wxWidgets/commit/440b8639708f57e4b66f4e83a0e8eafd6417b0c8, but I am unsure if it is strictly needed. It might fix some issue in OrcaSlicer though.

**Comment by liberodark** [comment]
Thanks for the information I will look at this this weekend and see what can be done.
For now I have spent a lot of time trying to correct it but I think I have not used the right method so I will approach orca-slicer directly moreover I know the software since I use it myself so I would like to correct it quickly for 24.11
Otherwise I will force libsoup2 to be used and return to webkitgtk_40

**Comment by liberodark** [comment]
Close for https://github.com/NixOS/nixpkgs/pull/360219

**Pull request closed without merging**